File: Database/networking/arp-poisoning/C#/arp-poisoning-tool/DatagramProvider.cs
using arp_poisoning_tool.DataStructures;
using arp_poisoning_tool.Packets;
using System.Runtime.InteropServices;
using System.Text.RegularExpressions;

namespace arp_poisoning_tool
{
    internal static class DatagramProvider
    {
        internal static Ethernet BuildEthernetPacket(string targetAddress, string sourceAddress, EthernetProtocolType type)
        {
            ValidatePhysicalAddress(targetAddress);
            ValidatePhysicalAddress(sourceAddress);

            return new Ethernet((ushort)type, 
                new MacAddress(ToByteArray(targetAddress.Split(':', '-'), 16)), 
                new MacAddress(ToByteArray(sourceAddress.Split(':', '-'), 16))
                );
        }

        internal static ArpReply BuildArpReplyPacket(string senderAddress, string senderIp, string targetAddress, string targetIp)
        {
            ValidatePhysicalAddress(targetAddress);
            ValidatePhysicalAddress(senderAddress);
            ValidateIpAddress(targetIp);
            ValidateIpAddress(senderIp);

            return new ArpReply(new MacAddress(ToByteArray(senderAddress.Split(':', '-'), 16)),
                new Ipv4Address(ToByteArray(senderIp.Split('.'))),
                new MacAddress(ToByteArray(targetAddress.Split(':', '-'), 16)),
                new Ipv4Address(ToByteArray(targetIp.Split('.')))
                );
        }

        internal static Ipv4 BuildIpv4Packet(InternetProtocolType protocol, string sourceIp, string destinationIp, int identification = 0, int ttl = 128, byte[] payload = null)
        {
            if (ttl > 255 || ttl < 1)
                throw new Exception("TTL - Time To Live should be a number between 1 and 255");

            if (identification > 65535)
                throw new Exception("Identification - 0 for null (new packet); 1-65535 for an existing value");

            int totalLength = Marshal.SizeOf<Ipv4>();
            if (payload != null)
                totalLength = totalLength + payload.Length;
            
            ushort ident = 0;
            if (identification == 0)
            {
                Random rnd = new Random();
                ident = (ushort)rnd.Next(65535);
            }
            else
                ident = (ushort)identification;

            ValidateIpAddress(sourceIp);
            ValidateIpAddress(destinationIp);

            
            Ipv4 packet = new Ipv4(ident.Swap(), (byte)ttl, (byte)protocol, 0,
                new Ipv4Address(ToByteArray(sourceIp.Split('.'))),
                new Ipv4Address(ToByteArray(destinationIp.Split('.'))),
                ((ushort)totalLength).Swap()
                );

            return packet;
        }

        internal static Loopback BuildLoopbackPacket(uint family = 2)
        {
            return new Loopback(family);
        }

        private static byte[] ToByteArray(string[] strings, int nBase = 10)
        {
            int size = strings.Length;

            byte[] bytes = new byte[size];
            for (int i = 0; i < size ; i++)
            {
                bytes[i] = Convert.ToByte(strings[i], nBase);
            }
            return bytes;
        }

        private static void ValidatePhysicalAddress(string physicalAddress)
        {
            string pattern = @"^([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})$";
            if (!Regex.IsMatch(physicalAddress, pattern))
            {
                throw new Exception("Invalid format. Please provide a mac address with the following notation: ff:ff:ff:ff:ff:ff or ff-ff-ff-ff-ff-ff");
            }
        }

        private static void ValidateIpAddress(string ipAddress)
        {
            string ipPattern = @"^((25[0-5]|(2[0-4]|1\d|[1-9]|)\d)\.?\b){4}$";
            if (!Regex.IsMatch(ipAddress, ipPattern))
            {
                throw new Exception("Invalid format. Please provide a ip address with the following notation: 10.10.10.10");
            }
        }
    }

    internal enum EthernetProtocolType : ushort
    {
        Unknown = 0x0,
        ARP = 0x0608,
    }

    internal enum InternetProtocolType : byte
    {
        TCP = 0x06,
        UDP = 0x11
    }
}

File: Database/networking/arp-poisoning/C++/arp-poisoning-tool/datagram-provider.cpp
#include "datagram-provider.hpp"
#include <regex>
#include "utilities.hpp"
#include <iostream>

void ValidateIpv4Address(const char* ipAddress)
{
	const char* pattern = "(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])";

	if (!std::regex_match(ipAddress, std::regex(pattern)))
	{
		std::cerr << "Invalid format. Please provide a ip address with the following notation: 10.10.10.10" << std::endl;
		exit(1);
	}
}

void ValidatePhysicalAddress(const char* physicalAddress)
{
	const char* pattern(
		"^([0-9A-Fa-f]{2}[:-]){5}"
		"([0-9A-Fa-f]{2})|([0-9a-"
		"fA-F]{4}\\.[0-9a-fA-F]"
		"{4}\\.[0-9a-fA-F]{4})$");

	if (!std::regex_match(physicalAddress, std::regex(pattern)))
	{
		std::cerr << "Invalid format.Please provide a mac address with the following notation : ff:ff:ff:ff:ff:ff or ff - ff - ff - ff - ff - ff" << std::endl;
		exit(1);
	}
}

MACADDRESS GetMACFromMACString(const char* str)
{
	MACADDRESS mac;
	int arrayIndex = 0;

	size_t size = strlen(str);
	std::string address(12, '0');
	int addressIndex = 0;
	for (int i = 0; i < size; i++)
	{
		if (isxdigit(str[i]))
		{
			address[addressIndex] = str[i];
			addressIndex++;
		}
	}

	for (int i = 0; i < 12; i += 2) {
		std::string byteString = address.substr(i, 2);
		BYTE byteValue = static_cast<BYTE>(
			stoi(byteString, nullptr, 16));

		mac.BYTES[arrayIndex] = byteValue;
		arrayIndex++;
	}

	return mac;
}

IPV4ADDRESS GetIPFromIPString(const char* str)
{
	IPV4ADDRESS ip;
	int index = 0;
	std::vector<BYTE> vector;

	size_t size = strlen(str);

	std::string address = str;
	size_t pos = 0;

	std::string delimiter = ".";
	std::string token;
	while ((pos = address.find(delimiter)) != std::string::npos) {
		token = address.substr(0, pos);

		BYTE byteValue = static_cast<BYTE>(
			stoi(token, nullptr, 10));

		ip.BYTES[index] = byteValue;
		address.erase(0, pos + delimiter.length());
		index++;
	}
	ip.BYTES[3] = static_cast<BYTE>(stoi(address, nullptr, 10));

	return ip;
}

PIPV4 CreateIpV4Packet(BYTE protocol, const char* sourceIp, const char* destinationIp, int identification, int ttl, BYTE* payload, int payloadSize)
{
	if (ttl > 255 || ttl < 1)
	{
		std::cerr << "TTL should be a value between 1 and 255" << std::endl;
		exit(1);
	}

	if (identification > 65535)
	{
		std::cerr << "Identification field should exceed 65535" << std::endl;
		exit(1);
	}

	int totalLength = sizeof(IPV4);
	if (payload != NULL)
		totalLength = totalLength + payloadSize;

	ValidateIpv4Address(sourceIp);
	ValidateIpv4Address(destinationIp);

	IPV4 packet;
	packet.Identification = 0;
	packet.TTL = ttl;
	packet.Protocol = protocol;
	packet.TotalLength = 0;

	int size = sizeof(packet);
	void* heap = malloc(size);
	
	if (heap == 0)
	{
		std::cerr << "Failed to allocate heap memory." << std::endl;
		exit(1);
	}

	memcpy(heap, &packet, size);

	return reinterpret_cast<PIPV4>(heap);
}

PETHERNET CreateEthernetPacket(USHORT type, const char* targetAddress, const char* sourceAddress)
{
	ValidatePhysicalAddress(targetAddress);
	ValidatePhysicalAddress(sourceAddress);

	MACADDRESS target = GetMACFromMACString(targetAddress);
	MACADDRESS source = GetMACFromMACString(sourceAddress);

	ETHERNET packet;
	packet.Type = type;

	packet.TargetAddress = target;
	packet.SourceAddress = source;

	int size = sizeof(packet);
	void* heap = malloc(size);

	if (heap == 0)
	{
		std::cerr << "Failed to allocate heap memory." << std::endl;
		exit(1);
	}

	memcpy(heap, &packet, size);

	return reinterpret_cast<PETHERNET>(heap);
}

PARPREPLY CreateArpReplyPacket(const char* senderAddress, const char* senderIp, const char* targetAddress, const char* targetIp)
{
	ValidatePhysicalAddress(targetAddress);
	ValidatePhysicalAddress(senderAddress);

	MACADDRESS bTarget = GetMACFromMACString(targetAddress);
	MACADDRESS bSender = GetMACFromMACString(senderAddress);

	IPV4ADDRESS bTargetIp = GetIPFromIPString(targetIp);
	IPV4ADDRESS bSenderIp = GetIPFromIPString(senderIp);

	ARPREPLY packet;

	packet.TargetAddress = bTarget;
	packet.SenderAddress = bSender;

	packet.TargetIp = bTargetIp;
	packet.SenderIp = bSenderIp;

	int size = sizeof(packet);
	void* heap = malloc(size);

	if (heap == 0)
	{
		std::cerr << "Failed to allocate heap memory." << std::endl;
		exit(1);
	}

	memcpy(heap, &packet, size);

	return reinterpret_cast<PARPREPLY>(heap);
}